google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java

```diff
@@ -0,0 +1,246 @@
+
+    // Inject the instrumented transport provider into a new client
+    BigtableDataSettings.Builder settingsBuilder =
+        testEnvRule.env().getDataClientSettings().toBuilder();
+
+    settingsBuilder
+        .stubSettings()
+        .setTransportChannelProvider(instrumentedTransportChannelProvider);
+
+    instrumentedClient = BigtableDataClient.create(settingsBuilder.build());
+  }
+
+  @After
+  public void teardown() {
+    if (instrumentedClient != null) {
+      instrumentedClient.close();
+    }
+    if (eventLoopGroup != null) {
+      eventLoopGroup.shutdownGracefully();
+    }
+  }
+
+  @Test
+  public void testFallback() throws InterruptedException {
+    // Verify that we can send a request
+    instrumentedClient.readRow(testEnvRule.env().getTableId(), "nonexistent-row");
+
+    // Enable the blackhole, which will prevent communication via IPv6 and thus DirectPath.
+    blackholeIPv6.set(true);
+
+    // Send a request, which should be routed over IPv4 and CFE.
+    instrumentedClient.readRow(testEnvRule.env().getTableId(), "nonexistent-row");
+
+    // Verify that the above check was meaningful, by verifying that the blackhole actually dropped
+    // packets.
+    assertThat(numBlocked.get()).isGreaterThan(0);
+
+    // Make sure that the client will start reading from IPv6 again by sending new requests and
+    // checking the injected IPv6 counter has been updated.
+    blackholeIPv6.set(false);
+    numIPv6Read.set(0);
+
+    // Arbitrary count that will ensure that there have been enough activity via IPv6.
+    final int MIN_COMPLETE_READ_CALLS = 40;
+
+    while (numIPv6Read.get() < MIN_COMPLETE_READ_CALLS) {
+      for (int i = 0; i < 20; i++) {
+        instrumentedClient.readRow(testEnvRule.env().getTableId(), "nonexistent-row");
+      }
+
+      Thread.sleep(100);
+    }
+
+    assertWithMessage("Failed to upgrade back to DirectPath")
+        .that(numIPv6Read.get())
+        .isGreaterThan(MIN_COMPLETE_READ_CALLS);
+  }
+
+  /**
+   * This is a giant hack to enable testing DirectPath CFE fallback.
+   *
+   * <p>It unwraps the {@link ComputeEngineChannelBuilder} to inject a NettyChannelHandler to signal
+   * IPv6 packet loss.
+   */
+  private void injectNettyChannelHandler(ManagedChannelBuilder<?> channelBuilder) {
+    try {
+      // Extract the delegate NettyChannelBuilder using reflection
+      Field delegateField = ComputeEngineChannelBuilder.class.getDeclaredField("delegate");
+      delegateField.setAccessible(true);
+
+      ComputeEngineChannelBuilder gceChannelBuilder =
+          ((ComputeEngineChannelBuilder) channelBuilder);
+      Object delegateChannelBuilder = delegateField.get(gceChannelBuilder);
+
+      NettyChannelBuilder nettyChannelBuilder = (NettyChannelBuilder) delegateChannelBuilder;
+      nettyChannelBuilder.channelFactory(channelFactory);
+      nettyChannelBuilder.eventLoopGroup(eventLoopGroup);
+    } catch (NoSuchFieldException | IllegalAccessException e) {
```

Comment:
can these exceptions simply bubble?

Reply:
No, this method gets invoked by a callback that doesn't allow checked exceptions